CoreAlgorithmPlugins/src/au/gov/asd/tac/constellation/plugins/algorithms/sna/similarity/CosineSimilarityPlugin.java

```diff
@@ -109,94 +109,128 @@ public void edit(final GraphWriteMethods graph, final PluginInteraction interact
-                    final float cosineSimilarity = neighboursMagnitude == 0 ? 0 : neighbourDotProduct / neighboursMagnitude;
-                    cosineSimilarities.put(Tuple.create(vertexOneId, vertexTwoId), cosineSimilarity);
                 }
             }
         }
-
-        // update the graph with cosine similarity values
-        SimilarityUtilities.addScoresToGraph(graph, cosineSimilarities, COSINE_SIMILARITY_ATTRIBUTE);
-
         // complete with schema
         PluginExecution.withPlugin(VisualSchemaPluginRegistry.COMPLETE_SCHEMA).executeNow(graph);
     }
-
-    private float dot(final int[] vectorOne, final int[] vectorTwo) {
-        assert vectorOne.length == vectorTwo.length;
-
-        float dot = 0;
-        for (int index = 0; index < vectorOne.length; index++) {
-            dot += vectorOne[index] * vectorTwo[index];
+    
+    private class VertexWithNeighbours {
+        private final int vertexId;
+        private final HashMap<Integer, Integer> neighbourWeightsMap;
+        private BitSet neighbours;
+        private final boolean selected;
+        private Float magnitude = null;
+
+        public VertexWithNeighbours(final int vertexId, final int vertexCount, final boolean selected) {
+            this.neighbourWeightsMap = new HashMap<>();
+            this.vertexId = vertexId;
+            this.selected = selected;
+            this.neighbours = new BitSet(vertexCount);
         }
-
-        return dot;
+           
+        private void addNeighbour(final int neighbourPosition, final int additionalWeight) {
+            int currentWeight = neighbourWeightsMap.getOrDefault(neighbourPosition, 0);
+            neighbourWeightsMap.put(neighbourPosition,currentWeight + additionalWeight);
+            neighbours.set(neighbourPosition, true);
+        }
+        
+        private float getMagnitude() {
+            if (magnitude == null) {
+                magnitude = calculateMagnitude();
+            }
+            return magnitude;
+        }
+        
+        private float calculateMagnitude() {
+            float mag = 0;
+            for(int neighbourWeight : neighbourWeightsMap.values()) {
+                mag += Math.pow(neighbourWeight, 2);
+            }
+            return (float) Math.sqrt(mag);
+        }
+        
     }
-
-    private float magnitude(final int[] vector) {
-        float magnitude = 0;
-        for (int index = 0; index < vector.length; index++) {
-            magnitude += Math.pow(vector[index], 2);
+    
+    BitSet getCommonNeighbours(final BitSet leftVertexNeighbours, final BitSet rightVertexNeighbours) {
+        final BitSet intersection = (BitSet) leftVertexNeighbours.clone();
+        intersection.and(rightVertexNeighbours);
+        return intersection;
+    }
+    
+    float getNeighbourDotProduct(final VertexWithNeighbours vertex1, final VertexWithNeighbours vertex2, BitSet intersection) {
+        float dot=0;
```

Comment:
minor but adding spacing between `dot`, `=`, and `0` would be good

Reply:
Done